### PR TITLE
feat: support existing scc instance

### PR DESCRIPTION
**Pull request by Soaib024**
### Description
https://github.com/terraform-ibm-modules/terraform-ibm-scc/issues/138

https://github.com/terraform-ibm-modules/terraform-ibm-scc-da/issues/140

https://github.com/terraform-ibm-modules/terraform-ibm-scc-da/pull/138#pullrequestreview-2153192310
Support existing scc instance

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Comment by Soaib024** [comment]
I have added data lookup for existing scc instance so that name of instance could be added in output as in we did for crn and region

File: variables.tf [hunk 1]
```diff
@@ -9,13 +9,19 @@ variable "resource_tags" {
...
+
 variable "instance_name" {
   type        = string
   description = "Name of the security and compliance instance that will be provisioned by this module"
 }
 
 variable "attach_wp_to_scc_instance" {
-  type        = string
+  type        = bool
   description = "When set to true, a value must be passed for the `wp_instance_crn` inout variable."
```
Comment (ocofaigh) [inline variables.tf:25]: just spotted a typo: `inout` -> `input`

File: outputs.tf [hunk 2]
```diff
@@ -1,35 +1,35 @@
...
+  value       = local.scc_instance_crn
+  # Don't return the SCC CRN until it has been configred with COS, as it can't be used until COS integration complete
   depends_on = [ibm_scc_instance_settings.scc_instance_settings]
 }
 
 output "name" {
-  description = "The name of the SCC instance created by this module"
-  value       = resource.ibm_resource_instance.scc_instance.name
+  description = "The name of the SCC instance."
+  value       = var.existing_scc_instance_crn == null ? resource.ibm_resource_instance.scc_instance[0].name : data.ibm_resource_instance.scc_instance[0].name
```
Comment (ocofaigh) [inline outputs.tf:24]: I don't see name here? https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/data-sources/resource_instance#attribute-reference

File: main.tf [hunk 3]
```diff
@@ -61,8 +76,9 @@ resource "time_sleep" "wait_for_scc_cos_authorization_policy" {
 
 # attach a COS bucket and an event notifications instance
 resource "ibm_scc_instance_settings" "scc_instance_settings" {
+  count       = var.existing_scc_instance_crn == null ? 1 : 0
```
Comment (ocofaigh) [inline main.tf:79]: this means that a user will be unable to an event notifications CRN to an existing SCC instance. I think we will want to support that use case, so you should probably use some dynamic blocks here to allow that to work

**Review by ocofaigh — CHANGES_REQUESTED** [review]
some comments

File: main.tf (hunk 3 above)
Comment (Soaib024) [reply main.tf]: Tried with dynamic blocks as follows:

```
  dynamic "event_notifications" {
    for_each = var.en_instance_crn != null ? { crn = var.en_instance_crn } : {}
    content {
      instance_crn = var.en_instance_crn
    }
  }
```

But this does not seem to work, as both the `event_notifications` and `object_storage` blocks are required:

```
 │ Error: Insufficient event_notifications blocks
│ 
│   on ../../main.tf line 83, in resource "ibm_scc_instance_settings" "scc_instance_settings":
│   83: resource "ibm_scc_instance_settings" "scc_instance_settings" {
│ 
│ At least 1 "event_notifications" block is required.
```

Instead, used a conditional operator which is working fine:

```
  event_notifications {
    instance_crn = var.en_instance_crn != null ? var.en_instance_crn : null
  }
  object_storage {
    instance_crn = var.cos_instance_crn != null ? var.cos_instance_crn : null
    bucket       = var.cos_bucket != null ? var.cos_bucket : null
  }
```

I also had to make `var.cos_instance_crn` and `var.cos_bucket` optional with the validation that either both have to be null or both have to be not null.

File: variables.tf (hunk 1 above)
Comment (Soaib024) [reply variables.tf]: addressed

File: outputs.tf (hunk 2 above)
Comment (Soaib024) [reply outputs.tf:24]: Screenshot 2024-07-04 at 8 31 54 PM

Name is available in data look up might be a documentation error

File: main.tf (hunk 3 above)
Comment (Soaib024) [reply main.tf]: @ocofaigh Okay, this will actually override the existing setting. I guess it might be a bad idea. I think we can add a new variable here.

**Comment by Soaib024** [comment]
@ocofaigh I have added a new variable `update_scc_settings`. I think this is still not a perfect solution because, for example, if I have an existing SCC instance with EN and COS configured, and I want to change the EN only, I will need to pass the COS CRN and bucket as well. If not, then the COS attachment will be removed if the COS CRN and bucket are null and vice-versa. Apart from this edge case, I think the solution works fine

**Comment by Soaib024** [comment]
> @ocofaigh I have added a new variable `update_scc_settings`. I think this is still not a perfect solution because, for example, if I have an existing SCC instance with EN and COS configured, and I want to change the EN only, I will need to pass the COS CRN and bucket as well. If not, then the COS attachment will be removed if the COS CRN and bucket are null and vice-versa. Apart from this edge case, I think the solution works fine.
> 
> I have also made var.cos_instance_crn and var.cos_bucket option

@ocofaigh do you have any comments on this one?

**Comment by ocofaigh** [comment]
@Soaib024 what if we broke it down into 2 variables such as `configure_cos_existing_instance` and `configure_en_existing_instance` ? That way you could dynamically update the settings based on these values?

**Comment by Soaib024** [comment]
> @Soaib024 what if we broke it down into 2 variables such as `configure_cos_existing_instance` and `configure_en_existing_instance` ? That way you could dynamically update the settings based on these values?

@ocofaigh 

We will still have a same problem because  both event_notifications  and  object_storage block are required here
```
resource "ibm_scc_instance_settings" "scc_instance_settings" {
  count       = var.existing_scc_instance_crn == null || var.update_scc_settings ? 1 : 0
  depends_on  = [time_sleep.wait_for_scc_cos_authorization_policy]
  instance_id = local.scc_instance_guid

  event_notifications {
    instance_crn = var.en_instance_crn
  }
  object_storage {
    instance_crn = var.cos_instance_crn
    bucket       = var.cos_bucket
  }
}
  ```
Screenshot 2024-07-16 at 5 59 55 PM

  
  So, when we try to set one of these we would be forced to update other as well.

**Comment by Soaib024** [comment]
Some possible solutions:

- No option to set the en and cos settings when an existing instance is used, so the user will have to manage these settings from where the instance was created.
- Either set both en and cos or none.
- Maybe raising a provider issue to make the en and cos block optional.

**Comment by ocofaigh** [comment]
yea but you can do this:

```
event_notifications {
    instance_crn = null
  }
```

**Comment by Soaib024** [comment]
> yea but you can do this:
> 
> ```
> event_notifications {
>     instance_crn = null
>   }
> ```

This will remove EN attachment if any, and this is the edge case that I mentioned [here](https://github.com/terraform-ibm-modules/terraform-ibm-scc/pull/121#issuecomment-2210678087)

**Comment by ocofaigh** [comment]
I was suggesting something like this:

```
event_notifications {
    instance_crn = var.existing_scc_instance_crn != null && var.configure_en_existing_instance ? var.en_instance_crn : null
  }
```

**Comment by ocofaigh** [comment]
Need to think about this.. leave it with me overnight ..

**Comment by Soaib024** [comment]
If we didn't find a solution that works with only condition then a workaround might be to pull the existing attachment with a data lookup. If the user wants to update only one of the attachments, update that one with the user input and update the other with the data from the lookup.

**Comment by ocofaigh** [comment]
@Soaib024 I think lets park this effort for now and maybe add to next deep dive agenda to see if this is something we want to support. For now, lets focus on adding existing SCC support to the DA itself - aka what you were doing in https://github.com/terraform-ibm-modules/terraform-ibm-scc-da/pull/138

File: main.tf (hunk 3 above)
Comment (in-1911) [reply main.tf]: @ocofaigh - there is a problem with using existing Event Notification instance with SCC.
It turns out that when the ibm_scc_instance_settings resource is used, it creates a source in the EN instance with the same name "compliance", so if two SCC instances are provisioned with this module, the second one trying to use EN will fails with the error about "source with the same name already exists. 

It probably has to be fixed at the provider level as there is no name parameter in [SCC instance settings resource](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/scc_instance_settings#argument-reference), but when the integration is created in SCC UI, the source name is set to SCC instance name.

Comment (Soaib024) [reply main.tf]: @ocofaigh @in-1911 I have removed the logic to set `scc_instance_settings` when using an existing scc instance for now, we could add it later once we have fix in provider

**Comment by Soaib024** [comment]
@ocofaigh Could you please review this PR,

The logic to update `scc_instance_settings` has been removed when existing scc instance is used, as a follow-up fix is needed in the provider to:
- [Expose the name parameter for EN settings](https://github.com/terraform-ibm-modules/terraform-ibm-scc/pull/121#discussion_r1706134296)
- [Also may be to allow updating COS and EN settings individually if possible](https://github.com/terraform-ibm-modules/terraform-ibm-scc/pull/121#issuecomment-2230770853)

**Review by ocofaigh — CHANGES_REQUESTED** [review]
I think the changes look good, however there is a test gap. I see you added `existing_scc_instance_crn` to the complete example, but none of the tests are using it. I suggest to add a test which will first provision an SCC instance (add some tf files to tests/resources like we do in other repos) and have a test provision that first, and then test the complete example passing a value for `existing_scc_instance_crn`.
For reference take a look at the existing resources test in the [terraform-ibm-scc-da](https://github.com/terraform-ibm-modules/terraform-ibm-scc-da/tree/main) repo

**Comment by Soaib024** [comment]
> I think the changes look good, however there is a test gap. I see you added `existing_scc_instance_crn` to the complete example, but none of the tests are using it. I suggest to add a test which will first provision an SCC instance (add some tf files to tests/resources like we do in other repos) and have a test provision that first, and then test the complete example passing a value for `existing_scc_instance_crn`. For reference take a look at the existing resources test in the [terraform-ibm-scc-da](https://github.com/terraform-ibm-modules/terraform-ibm-scc-da/tree/main) repo

I've added a new test to pass an existing SCC instance to the complete example, but I'm wondering if we should remove the complete test since we already have a basic test that creates an SCC instance with COS settings. I guess we won't introduce test gaps removing it.

**Comment by Soaib024** [comment]
@ocofaigh I have moved the completeExampleTest to other_test.go

File: tests/resources/existing-resources/main.tf [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  cos_instance_name      = "${var.prefix}-cos"
+  kms_encryption_enabled = false
+  retention_enabled      = false
+  resource_group_id      = module.resource_group.resource_group_id
+  bucket_name            = "${var.prefix}-cb"
+}
+
+module "create_scc_instance" {
+  source                            = "terraform-ibm-modules/scc/ibm"
+  version                           = "1.7.2"
```
Comment (ocofaigh) [inline tests/resources/existing-resources/main.tf:20]: SCC module is local to this repo, so you should use a local source. Please be aware that when you do that, you may need to update the test itself, same way we did in https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/pull/117

**Review by ocofaigh — CHANGES_REQUESTED** [review]
1 minor comment

File: tests/resources/existing-resources/main.tf (hunk 4 above)
Comment (Soaib024) [reply tests/resources/existing-resources/main.tf]: addressed

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**